### DISCUSSION TITLE
Has anyone tried MikroORM in a monorepo?

### DISCUSSION THREAD

**Discussion by AlexandrosMelissas** [mikro-orm/mikro-orm#4075] [Q&A] [answered]
Hello. I am trying to make my monorepo work with MikroORM. My current monorepo consists of only 3 apps (1 backend with mikro and 2 frontends) and some shared libraries. My ultimate goal is to be able share some stuff like validation and types across all apps. The only problem i have encountered so far is with MikroORM unique entity typings like Collections and Identified references. I am trying to make this code work:

```
@Entity()
export class Host implements HostType {
  @PrimaryKey()
  id!: string;

  @Property()
  firstName!: string;

  @Property()
  lastName!: string;

  @Property()
  email!: string;

  @OneToMany(() => House, 'host')
  houses = new Collection<House>(this);
}
```

Here we have entity Host which implements HostType ( which is coming from the shared @app-name/types library i have made). The problem is that the types library itself is not aware of collection types or identified references because it has to be used from frontend as well & it needs to be independent. Having different types for front end and back is a no go because i am using a trpc-like library to reuse the types for endpoint responses. [Ts-Rest Library](https://github.com/ts-rest/ts-rest) Maybe there is another way of doing it? Im really interested to see if anyone has managed to do this in a monorepo.

**Comment by B4nan** [comment, marked as answer]
> Having different types for front end and back is a no go because i am using a trpc-like library to reuse the types for endpoint responses.

This seems like the wrong line of thinking. The entity definition is how the backend works with it, it is not the same as the JSON that gets transferred via network - which is what your interfaces that trpc uses are. Don't try to make it the same.

You can use types like `EntityDTO` and `EntityData` to convert entity type to a DTO/POJO type - those don't have collection or reference wrappers. Search the issues and discussions, there was a question about those types some time ago.

**Comment by AlexandrosMelissas** [comment]
I guess you are right. Even if Mikro didn't have special entity types, it would still be considered a "bad" practice, right? I made it work by wrapping the entity and then POJO it before i return it as a response. Thanks for pointing me to the right direction.

**Reply by B4nan** [reply]
You would have the same issue with any relationship - imagine a book that has author, on frontend it might be a FK or populated entity, and maybe that populated entity will have some other relations populated too. This is all abstracted in the entity definition, but you need to handle the difference on frontend somehow.
